### DISCUSSION TITLE
Triggerer memory gradually increases to 1GB mark and then CPU spins up to 100% and it would no longer sends heartbeat

### DISCUSSION THREAD

**Discussion by omarsmak** [apache/airflow#38671] [Q&A] [answered]
Hello folks!

We have been wrestling with a very "weird" issue, where we notice that triggerer memory gradually increases to 1GB over few hours window, when it reaches the ~1GB mark, the CPU starts to spin to 100% and the triggerer becomes unresponsive and no longer sending heartbeat messages which ultimately we will have to restart the pod manually. This chart can show the co-relation here:
 
<img width="1691" alt="Screenshot 2024-04-01 at 10 49 25" src="https://github.com/apache/airflow/assets/1190920/05c5cad9-d99b-4c60-b8e3-9020327468c1">

Taking the learning from https://github.com/apache/airflow/issues/14924, definatly we are using  `container_memory_working_set_bytes` to report the correct memory metrics. Also, we had a suspicious that might be related to the page cache memory but looking at the "dirty" memory in the pod, it looks to me is correctly flushed (which indicates no leaks):
```
airflow@triggerer-7cbcc464-vdtcr:/opt/airflow$ cat /proc/meminfo | grep "Dirty"
Dirty:               204 kB 
```

We tried to increase the pod memory request and limits to 4gb of memory but that didn't help, "weirdly" it always stops increasing at 1gb which affects the CPU to spin to 100%.

The the following is how we operate Airflow:
- We are deploying Airflow 2.8.1 on K8s using the official helm charts in GCP.
- We are using managed ClousSQL instance (PSQL) for the database, we tried with both, having Airflow directly talk to CloudSQL proxy vs via PGBouncer (well PGBouncer is far better for the performance in general), same results.
- We are trying to run ~100 concurrent DAGs, the DAG itself is pretty basic as it has several tasks that call other services internally via gRPC, this relies "heavily" on Async sensor and deferrable operators (which means triggerer is critical for us), as we noticed deferrable operators really helps to reduces the load on the worker with a good margin. Each Dag may have around thousands of DAGRuns in the queue.
- Our aim to reach ~10_000 concurrent dags to run, however, the triggerer issues is the bottleneck ATM.
- We have given the triggerer 4GB of memory and 4cores of CPU but that didn't help much to mitigate the issue.

Any other tips to troubleshoot this issue? It is highly apperciated to get some pointers here as we have wrestling a lot with this issue for the past week and we couldn't overcome it :(.

Thanks!
Omar Al-Safi

**Comment by potiuk** [comment]
First of all - I think it woud be helpful if you explain which Airflow version you are on and what is the deployment you use (K8S we know, Postgres too, but also the versions of those, whether you are using a chart (which ?) to deploy it etc. 

It would be great to see if  the "trigger" table gets bigger while you get more memory used (i.e. if there are more triggers waiting in the queue). It could also help if together with such analysis you could post which triggers you are using (and maybe if you see growth in the Trigger table - which ones are the ones that are growing in number.

Maybe a good idea would be to run `pyspy` on triggerer when it reaches 100% CPU - that might give us a clue where triggerer gets into - apparently some active loop. If you can catch triggerer "red-handed" in the high CPU loop, it might be useful to get it. Similarly. before it reaches the CPU - putting a `memray` (released by bloomberg) on running triggerer process might help to see what is leaking the memory. Likely you could then make a few dumps and compare them and we could see what is leaking the memory.

Having those memray/pyspy dumps / flamegraphs posted somewhere where we can access them might be really helpful to diagnose things.

It might be also the reason why some users are reporting that triggerer might restart frequently https://github.com/apache/airflow/issues/33647 

cc: @kaxil and the team - that might be something that we should look closely if we get more data.

**Reply by potiuk** [reply]
> We are deploying Airflow 2.8.1 on K8s using the official helm charts in GCP.

Missed that one :facepalm:

**Comment by omarsmak** [comment]
Thanks @potiuk !, here are the answers:

- Airflow is `2.8.1` (you noticed that ;))
- Airflow Helm charts version `1.11.0`.
- Postgress version `13.13`.
- K8s version `v1.27.8-gke.1067004`


I have been running `select count(*) from trigger`, looks to it doesn't increase or anything, it just keeps changing depending on the queue like:
91 -> 100 -> 80 -> 75 -> 90 -> 102 .. etc. Therefore, I don't see any issues here?

I will get you the memory dumps once I am able to access. 

Thank you!

**Reply by potiuk** [reply]
> 91 -> 100 -> 80 -> 75 -> 90 -> 102 .. etc. Therefore, I don't see any issues here?

Yeah. That's expected

**Comment by omarsmak** [comment]
Also in regards to the triggerer table in https://github.com/apache/airflow/issues/33647, I see the following the oldest entries in the `created_date` which were created today:

![Screenshot 2024-04-02 at 12 59 39](https://github.com/apache/airflow/assets/1190920/f639362e-aa66-4d4c-b348-5a5f3cbad313)

**Comment by omarsmak** [comment, marked as answer]
Hi @potiuk , 

Just to update you, after we have taken the heap dumps and analysed everything closely, since we use Opentelemetry Python SDK to capture the traces of gRPC within the DAG definition, turned out the Opentelemetry Python SDK was creating hundreds of tracing objects for every gRPC calls we do (and we do hundreds of these per minute) and that somehow contributed to the fast phase of growing memory plus this unexplained CPU spikes that made the triggerer becomes unhealthy. Once we removed the otel logic, triggerer went back to the "normal" operation mode where things are in the green zone. So at least now we know what is the cause and we can fix it :).

I thank you again for your time and tips, highly appreciated!   

Regards,
Omar

**Reply by potiuk** [reply]
Ah . Fantastic! thank you. This is an important finding as we are moving into more of the opentelemetry integration - not as much as you do for gRPC but tracing memory used by opentelemetry might be important thing (cc: @howardyoo @ferruzzi )
